Add RUN_OUTPUT to many runnable tests (first batch)

MoonlightSentinel opened the pull request:
These are mostly trivial outputs,  I'll add the more complex ones in further PR(s) (to keep them at reasonable size).

Furthermore some of these `printf` could probably be removed but that's a discussion for another ~~day~~ PR.

MoonlightSentinel committed:
Add RUN_OUTPUT to many runnable tests (first batch)

And remove a non-standard `printf` from testaa2.d/mixin1.d which does
not work on Azure (it's content is verified by an assert anyway)

Geod24 commented:
Where is `RUN_OUTPUT` defined in the test runner ? Also, why do we have `TEST_OUTPUT:` (a section) but `RUN_OUTPUT` (a token?).

MoonlightSentinel commented:
> Where is `RUN_OUTPUT` defined in the test runner ?

Added in #10850 but I forgot to add it to the README. Thanks for the reminder!

> Also, why do we have `TEST_OUTPUT:` (a section) but `RUN_OUTPUT` (a token?).

Fixed. That was a typo but the colon was never enforced in the first place (it's just a convention AFAICT)

Geod24 commented on test/runnable/test52.d:
```diff
@@ -1,4 +1,11 @@
+RUN_OUTPUT:
+---
+this()
+~this()
```
CI not happy with this (missing the printf from module dtor, although why you'd do that is beyond me.

MoonlightSentinel replied:
It's checking that the module dtor is **not** run because X.~this should've aborted the test.

Geod24 commented:
Cool! And thanks for all the work you're doing on the CI. Much appreciated (and needed).

MoonlightSentinel commented on test/runnable/testaa2.d:
```diff
@@ -69,7 +81,6 @@ void foo2()
 void testaa()
 {
     size_t i = foo("abc");
-    printf("i = %zd\n", i);
```
Removed this `printf` because `z` (`size_t`) is not supported on Azure-mingw (and the assert below verifies i anway)

MoonlightSentinel committed:
Rework test52.d to avoid undefined behaviour

MoonlightSentinel commented:
Reworked `runnable/test52.d` to avoid `exit`

MoonlightSentinel committed:
Rework testgc2.d to achieve deterministic output

Geod24 commented on test/runnable/circular.d:
```diff
@@ -1,7 +1,14 @@
+
```
```suggestion
```

Geod24 commented on test/runnable/circular.d:
```diff
@@ -1,7 +1,14 @@
 // REQUIRED_ARGS: -d
 // PERMUTE_ARGS: -dw
 // EXTRA_SOURCES: imports/circularA.d
 // This bug is typedef-specific.
```
Ahem maybe this test is useless now...

MoonlightSentinel replied:
Probably. Not sure why it's a runnable test anyway because it targets a compiler error

MoonlightSentinel replied:
But  pruning the test suite is something for another PR IMHO.

Geod24 commented on test/runnable/testgc3.d:
```diff
@@ -1,5 +1,11 @@
 // PERMUTE_ARGS:
 // REQUIRED_ARGS:
+/*
```
Would be nice to merge  this  with the previous comments (`//`). Here and everywhere else.

MoonlightSentinel replied:
Can do, but that would be a very tedious endeavor ...

wilzbach commented on test/runnable/test52.d:
```diff
@@ -2,28 +2,26 @@
 {
-    void exit(int);
-    int printf(const char*, ...);
+    assert(count == 3);
```
Why not print the count here? After all this test is about destructors, so it shouldn't rely on them working to ensure test correctness, but if you print the count here the test suite can ensure that it was properly called.

MoonlightSentinel replied:
Makes sense, will add a printf.

wilzbach reviewed (approved):
Awesome work! This is a huge improvement to the test suite!
BTW does RUN_OUTPUT already work with AUTO_UPDATE already?

MoonlightSentinel commented:
> BTW does RUN_OUTPUT already work with AUTO_UPDATE already?

Yes, `d_do_test.d` throws a `CompareException` if a missmatch occurs which is then handled as usual.

thewilsonator commented:
Is this good to go?

MoonlightSentinel commented:
Think so but @Geod24 might want to have another look
